code/modules/reagents/chemistry/holder.dm

```diff
@@ -898,25 +898,107 @@
+			(check defines at `code\__DEFINES\reagents.dm`)
+		* blacklist_flag
+			the method will remove random reagents from the possible list when they have this flag. default NONE(0)
+			same rule above
+			(uses chemical defines)
+		* union
+			default TRUE. if FALSE, the same item will be added to the possible list, making some reagent higher chance to spawn when a reagent is called more than once.
+			you will hardly use this though.
+			(Bicaridine, Bicardine, Bicaridine means 3x chance than normal.)
+
+	 *--- How to add a new random reagent category ---*
+		1. add a new flag at 'code\__DEFINES\reagents.dm' and `var/list/chem_defines` below
+			i.e.) `#define CHEMICAL_SOMETHING_NEW (1<10)`
+		2. add a new static variable which is corresponding to the new flag.
+			i.e.) `var/static/list/random_reagents_xx = list() // CHEMICAL_SOMETHING_NEW`
+		3. add the new static variable to the 'random_reagent' list
+			then done! (of course, don't forget to turn on the new flag at each desired reagent)
+	*/
+
+
+	// ----below is a section you might want to edit for more chem RNGs----
+	var/static/list/chem_defines = list( // check `code/__DEFINES/reagents.dm`
+		CHEMICAL_NOT_SYNTH,     // (1<<0)
+		CHEMICAL_BASIC_ELEMENT, // (1<<1)
+		CHEMICAL_BASIC_DRINK,   // (1<<2)
+		CHEMICAL_RNG_GENERAL,   // (1<<3)
+		CHEMICAL_RNG_FUN,       // (1<<4)
+		CHEMICAL_RNG_BOTANY,    // (1<<5)
+		CHEMICAL_GOAL_CHEMIST_DRUG,         // (1<<20) - goal_define starts at 20.
+		CHEMICAL_GOAL_CHEMIST_BLOODSTREAM,  // (1<<21)
+		CHEMICAL_GOAL_BOTANIST_HARVEST,     // (1<<22)
+		CHEMICAL_GOAL_BARTENDER_SERVING)    // (1<<23)
+	var/static/list/random_reagents_a = list()  // CHEMICAL_NOT_SYNTH
+	var/static/list/random_reagents_b = list()  // CHEMICAL_BASIC_ELEMENT
+	var/static/list/random_reagents_c = list()  // CHEMICAL_BASIC_DRINK
+	var/static/list/random_reagents_d = list()  // CHEMICAL_RNG_GENERAL
+	var/static/list/random_reagents_e = list()  // CHEMICAL_RNG_FUN
+	var/static/list/random_reagents_f = list()  // CHEMICAL_RNG_BOTANY
+	var/static/list/random_reagents_goal_a = list()  // CHEMICAL_GOAL_CHEMIST_DRUG
+	var/static/list/random_reagents_goal_b = list()  // CHEMICAL_GOAL_CHEMIST_BLOODSTREAM
+	var/static/list/random_reagents_goal_c = list()  // CHEMICAL_GOAL_BOTANIST_HARVEST
+	var/static/list/random_reagents_goal_d = list()  // CHEMICAL_GOAL_BARTENDER_SERVING
+	var/static/list/random_reagent = list(
+		random_reagents_a,
+		random_reagents_b,
+		random_reagents_c,
+		random_reagents_d,
+		random_reagents_e,
+		random_reagents_f,
+		random_reagents_goal_a,
+		random_reagents_goal_b,
+		random_reagents_goal_c,
+		random_reagents_goal_d)
+	// ----above is a section you might want to edit for more chem RNGs----
+
+	// initialize random reagent static lists
+	if(!random_reagents_a.len)
+		var/i = 0
+		for(var/each_define in chem_defines)
+			i += 1
+			for(var/thing in subtypesof(/datum/reagent))
```

Comment:
I would think that subtypes of has quite a bit of overhead to it, so I think that it should be done on the outer for loop, so subtypesof is only called once, rather than chem_defines.len times